### ISSUE TITLE
pre-render phase

### ISSUE THREAD

**Issue by tanner0101** [vapor/leaf-kit#35]
Templates will often contain references to static variables that only need to be serialized once. For example:

```mustache
<link href="{{baseURL}}/style.css">
```

Currently TemplateKit offers no distinction between dynamic and static variables. There's potential for huge performance gains if TemplateKit could support a "pre-render phase" where variables known to be static are converted to raw data. 

There are a couple of ways this could be implemented. 

### Special AST Case

A new syntax type could be created that represents a "static" variable (as opposed to the normal variable type). When rendering an uncached view, TemplateKit could first serialize all instances of this static variable. 

In Leaf, this might look something like:

```html
<link href="#static(baseURL)/style.css">
Hello, #(name)!
```

Some caveats with this method are that the data would still need to be supplied in the render context even after it has been cached. This could also require some complex logic to "simplify" ASTs  after components have been replaced. 

### Separate Renderer

Another option is that we create a new type of "pre-renderer". This would take a different parser than the main renderer and would be run as a step before the main renderer. 

The pipeline for this would look like:

```
ViewData + StaticContext -> Pre-Renderer = ProcessedViewData
ProcessedViewData + Context -> Renderer = FinalViewData
```

This has the benefit that we can use an entirely different parser for the pre-process phase.

In Leaf, this might look something like:

```html
<link href="$(baseURL)/style.css">
Hello, #(name)!
```

The pre-processing parser would use `$` as it's special token, where as the normal parser would continue to use `#`. 

/cc @mcdappdev @BrettRToomey

**Comment by tdotclare** [comment]
Bumping this issue as I was thinking about implementing something similar in order to be able to have more application-agnostic templates, and part of why I did the LeafAST PR is that I think the caching system needs to be able to store *unresolved* documents in a pre-render phase so that things like static variables can handled as an extension of the current pipeline -

eg, runtime automatically registers global/static variables as *templates*, so that modifying the variable value itself automatically cascades to the referencing templates.

This would be particularly useful for MASSIVELY simplifying cases where you're serving multiple domains using the same handlers, for instance, so that you can inline the server_name the request is being processed without forcing you to refer specifically to domain-scoped templates

I'd suggest that rather than adding a new tag though, scoping rules be established - eg something like the below:
- **`#(key)`** - a context value for "key" (must be passed in context and annoying to handle when doing routes
- **`#($context.key)`** - as above, but clearly scoped
- **`#($global.key)`** - a scoped value for "key" in global static variables
- **`#($route.key)`** - a scoped value for "key" in per-route static variables- **`#($exports.key)`** - a scoped value for "key" in exported values (identical to `#import("key")`)
- **`#($exports.key, value)`** - a scoped value for "key" in exported values (identical to `#export("key")`)

## **Side note:**
Personally I find it confusing that there's not consistency between "is it a string literal or not" for parameters when it's context vs a variable (and also that there's some ambiguity in nested templates over how export<->import relates if one attempted to override exports). I do think it would clarify things to establish that #() represents ANY kind of variable, whether it be context data or a global/otherwise settable value, and that the default behavior is context data if not directly scoped elsewhere.

Something like the above would remove the need for unique #export/#import tags, and would also make it easier for handling **outside** of Leaf itself, so that middleware in Vapor could automatically set relevant variables like server name or requested path or whatever.

**Comment by tdotclare** [comment]
Massively Closed in LeafKit-1.0.0tau
